### ISSUE TITLE
[Bug]: Use 'default' highlights

### ISSUE THREAD

**Issue by rebelot** [akinsho/bufferline.nvim#320]
### What happened?

hello, I'm a colorscheme creator, but at the moment it is impossible to setup bufferline colors from highlights definitions unless the colorscheme is loaded _after_ bufferline, which is not really convenient. Would it be possible for you to set only `default` highlights so that they are not modified by bufferline if the colorscheme sets them beforehand?

### What did you expect to happen?

-

### Additional Information

-

**Comment by akinsho** [comment]
Hi @rebelot,

so this is kind of by design that this plugin isn't very easy to override via colorschemes. I actually created this plugin because I wanted to have a consistent bufferline UI that automatically adapted to whatever colorscheme I chose to use. I also wanted the appearance to be uniform across colorschemes. One thing I wanted to avoid was one colorscheme maker, for example deciding to make the tab foregrounds bright green or just tweaking the highlights in some unforeseen way. I essentially didn't envision nor try to make it possible for colorscheme authors to override these highlights because I wanted to make them opinionated and consistent across colorschemes and give only the *user* the ability to change highlights that they wanted. What I would really like to avoid is someone loading in this plugin and just seeing some completely different UI/highlighting to what I designed this plugin to do. Or for them to change colorscheme only for bufferline to suddenly have some radically different appearance.

I guess I need to document some mechanism for allowing colorschemes to maybe change highlights, but I wouldn't want that to be automatic, I would want a user to specifically opt into using the custom colouring and not have colorschemes change things by default.

**Comment by rebelot** [comment]
It would at least be nice to know exactly which default highlight groups are required by bufferline to work. I am opening this issue because I am updating [kanagawa](https://github.com/rebelot/kanagawa.nvim) to use the new `nvim_set_hl` api and this broke bufferline separators color. I imagine this is due to the fact that, at the moment, the new highlight api replaces the highlight fields instead of updating them and I am probably not setting some of the required base hl-groups the way bufferline expects.

**Comment by adrian5** [comment]
I ran into the same issue while playing with the new `nvim_set_hl` function:

```lua
hl(0, "Normal", { fg = "#d8dee9", bg = "#1b2b34" })
--                      ^^^^^^^ -- bufferline.nvim uses this exact color for the separator

-- :verbose hi Normal
-- Normal         xxx guifg=#d8dee9 guibg=#1b2b34
```

The old approach gives correct highlights for all elements of the tab bar:

```lua
vim.cmd("highlight Normal guifg=#d8dee9 guibg=#1b2b34")

-- :verbose hi Normal
-- Normal         xxx guifg=#d8dee9 guibg=#1b2b34
```
Maybe something about the way properties are applied in nvim, as @rebelot alluded to?

Would be prudent to find a solution to this before nvim-0.7 is released.

**Comment by akinsho** [comment]
@rebelot I've created #329 which allows setting highlights to be default if the `themable` is true in a user's config. But I've also clarified the base highlight groups bufferline expects to be set, and it is generally more advisable to set those and allow bufferline to do its own dynamic highlighting than trying to override the dynamic highlight directly, since as I mentioned that is not how I designed things to work as I might choose to change the highlight name at some point but keep the keys in the user table the same, so it won't break for users but would for anyone depending on the highlight names specifically.

Regarding `nvim_set_hl` I believe this issue is not unique to bufferline. I recently migrated to using it in my config and realized that it doesn't inherit the highlight fields of an existing highlight. In this case, I think it's best to fetch the hl by name and then do a `tbl_extend`.

Anyway, please have a go with it and let me know if it allows you to solve this issue.

**Comment by akinsho** [comment]
@adrian5 thanks for the heads-up, I've been tracking its release, although tbh I'm not sure how any of what is to come requires some urgent change in this plugin. The issue in https://github.com/akinsho/bufferline.nvim/issues/320#issuecomment-1030673058 is less to do with something this plugin does and more to do with the fact that any highlights now set via the API obliterate the existing hl rather than extend it. I've moved most of my personal config highlights to use `nvim_set_hl` and I think almost all plugin authors will probably want to use a mechanism where they fetch the existing highlight set the new value on it and update the whole thing, so it does an update rather than a replacement.

All the PR I linked really does is add some docs and make it possible to have the highlights be default, but that's opt in not opt out, so not very much will change. At least, hopefully the docs will make it clearer to colorscheme authors what groups need to be set.

**Comment by akinsho** [comment]
> The highlight groups you listed in your PR are fully set here, unless I'm missing something.

Not sure exactly what you meant here, but these are neovim native highlights, none of them are highlights created by this plugin. The entire way that this plugin works is by finding native highlights that have a certain look (or should have a certain look) regardless of colorscheme i.e. all colorschemes should set these, and they try and work similarly for each theme and using them as a base for all it's highlighting.

**Comment by atchim** [comment]
I also had been trying to add support for bufferline until I realized that the plugin has an autocommand for the `ColorScheme` event which overrides the previous highlight definitions set by the color scheme. I don't think that's a good idea that the plugin automatically generates highlight definitions based on builtin groups previously set by the color scheme. In general, the bufferline doesn't look great and, in some cases, is even hard to differ between tabs based on its colors.

In the current situation, while #329 is not merged, the user can force bufferline to use certain highlight definitions through the `highlights` field in the setup table, which will cause bufferline to use the same colors even after changing the color scheme. A workaround for this inflexibility is to redefine the highlight groups right after bufferline redefine the groups previously set by a color scheme. This situation leads to a lot of redefinition of highlight groups, which I think is annoying.

Anyway, thanks for this plugin! I enjoy it so much!

**Comment by akinsho** [comment]
@atchim there are quite a few different things raised in the statement above, and tbh it's not entirely clear if your issue strictly relates to what this issue here is about. I really want this issue to be restricted to just colorscheme authors and issues re. theming (where possible) this plugin.

> I realized that the plugin has an autocommand for the ColorScheme event which overrides the previous highlight definitions set by the color scheme

Re-applying colours which a plugin created on the `ColorScheme` autocommand is far from "strange" or incorrect, since in nvim if a plugin is creating highlights this is the only way to ensure they survive resourcing your vim config or changing themes.

> I don't think that's a good idea that the plugin automatically generates highlight definitions based on builtin groups previously set by the color scheme. 

So there might be a bug being described here, i.e. the colours don't change when the colorscheme changes? Because they are supposed to change with the colorscheme and that may be broken. I think the central premise though might be completely against how this plugin was designed, *the entire point of dynamic highlighting in this plugin* is that it takes a consistent group of colours as a baseline palette and then can lighten or darken them to produce the results it does. It's kind of a core aspect of what this plugin does, it doesn't work to leave that to colorscheme authors since there's no chance the desired UI would be produced. Also, as I've mentioned many times, I envisioned this plugin specifically as working independently of colorschemes specifically since I wanted it to be consistent regardless of colorscheme.

> In general, the bufferline doesn't look great and, in some cases, is even hard to differ between tabs based on its colors.

It's not very clear to me what this comment means? i.e. the out-of-the-box experience as I envisioned is not pleasant to you, in which case that is actually fine as there are a bunch of alternatives to this plugin. The out-of-the-box experience is broken for you, and that is what you are reporting? As I mention in the README this plugin was not designed to fit all colorschemes. If you have specific examples of what "doesn't look great" looks like, that would be helpful

> the user can force bufferline to use certain highlight definitions through the highlights field in the setup table,

Again this is not a hack/"forcing" but literally how this was designed to be used.

> which will cause bufferline to use the same colors even after changing the color scheme

Bufferline is supposed to change colours when the colorscheme changes, if you pass the setup function a red colour though and then change theme, and it's still red that's really no this plugins fault

@atchim it seems that you might be experiencing issues that aren't directly related to this issue and more along the lines of issues with the initial UI of bufferline requiring it to be tweaked. It's probably best to raise a separate issue for those, the entire premise behind bufferline (granted it is opinionated) is that it shouldn't need to be tweaked by anyone, I appreciate that's basically impossible in the nvim community of tinkerers, but it's supposed to be as hands off as possible.

**Comment by adrian5** [comment]
> The issue in [#320 (comment)](https://github.com/akinsho/bufferline.nvim/issues/320#issuecomment-1030673058) is less to do with something this plugin does and more to do with the fact that any highlights now set via the API obliterate the existing hl rather than extend it.

I'm probably misunderstanding something about the causal chain of events here. Are you saying that running `nvim_set_hl` leads to decimated hl-Groups before the plugin sources them, or that setting them _afterwards_ breaks the already set up plugin UI?

The issue I described in my [earlier comment](https://github.com/akinsho/bufferline.nvim/issues/320#issuecomment-1042491849) is about setting highlights before loading the plugin. And since these highlights look "complete" to me (when queried via `:highlight Group`, I'm wondering what it is about `nvim_set_hl` vs `:highlight` that causes the plugin to have issues.

```lua
vim.cmd("highlight Normal guifg=#c5ccd7 guibg=#1b2b34 ctermfg=252 ctermbg=235")
```

![normal](https://user-images.githubusercontent.com/5426924/162190097-9f22c06c-fdad-4d33-b440-346c2b795886.png) :+1: 


```lua
nvim_set_hl(0, "Normal", { fg = "#c5ccd7", bg = "#1b2b34", ctermfg = 252, ctermbg = 235 }) 
```

![broken](https://user-images.githubusercontent.com/5426924/162190550-322409e2-8dd5-4d18-8213-f9ca2879d919.png) :broken_heart:

**Comment by akinsho** [comment]
@adrian5 thanks for clarifying with the commands and images, that helps *a lot*. So my comment about `nvim_set_hl` was more that if you run `hi Group guifg=Red` for example the underlying `Group` which might actually by default be `Group guifg=blue gui=italic` etc. will now be `Group guifg=Red gui=italic` in the native nvim command whereas with `nvim_set_hl` it's almost as if you ran `hi clear Group` and then ran your highlight command i.e. it completely unsets the group you call `nvim_set_hl` on.

I believe this is rarely what colorscheme authors expect after years of it working differently with the old `:highlight` but that's a choice that up to the maintainers 🤷🏿‍♂️. So my thought was that missing highlights here are due to that, i.e. the other attributes being removed during the migration to `nvim_set_hl` which is something colorscheme authors now need to work around on their side.

Regarding your problem, as far as I can see this is possibly a bug in nvim or maybe some sort of new obscure usage I'm not sure.

So if you set
```lua
vim.cmd("highlight Normal guifg=#c5ccd7 guibg=#1b2b34 ctermfg=252 ctermbg=235")
```
Then restart neovim and run
```
lua vim.pretty_print(vim.api.nvim_get_hl_by_name('Normal', true))
```
*Then* repeat the process above with the `nvim_set_hl` call, you'll notice that the returned value from `nvim_get_hl_by_name` is different. I don't know why that would be, this plugin uses the API to query the values of the groups I mentioned above, so this issue will be because the values are incorrect coming back from the API for some reason (If you can reproduce this that is)

**Comment by adrian5** [comment]
Right, that is weird:

```lua
-- :highlight
{                                                                                                                                                                                                                                     
  background = 1780532,
  foreground = 12963031
}

-- nvim_set_hl
{                                                                                                                                                                                                                                    
  [true] = 6
}
```

For the time being, what I do in my colorscheme is set the `Normal` group the old way, and all other groups via `nvim_set_hl`. This way bufferline looks as expected. But of course I'd like to get rid of that exception at some point.

**Comment by atchim** [comment]
@akinsho, I never did a plugin which defines highlight groups for itself before, so I didn't know that plugins have to define its groups every time the color scheme changes. So, my bad for complaining about this. Although, the [Leap](https://github.com/ggandor/leap.nvim) plugin does have a nice approach that is the definition of an `User` event with the `LeapEnter` pattern. This way, the user can create autocommands for redefining the highlight groups, for example.

**Comment by akinsho** [comment]
@atchim there is no need for a custom autocommand from this plugin because

a.) you can set up your own colorscheme autocommand to re-apply your highlight overrides, and this is generally good practice anyway whether you use bufferline if you are overriding highlights in your config it's a good idea to do that.
b.) if you set highlights or overrides via this plugin, then there's no need to worry about them being overridden by this plugin 🤷🏿‍♂️

**Comment by akinsho** [comment]
@adrian5 I think all of this is worth reporting upstream since it seems like something is broken on the API side, all this is fairly new so wouldn't surprise me if there were just a few random bugs, also from randomly skimming neovim issues I think the highlighting for `Normal` is actually done in a convoluted way that was inherited from vim so think that could be why that specific highlight is behaving weirdly

**Comment by akinsho** [comment]
@adrian5 thanks for raising that issue with nvim 👍🏿 hopefully when that's fixed that aspect of the problem should be resolved.

Regarding this issue as it was initially raised about having default highlights, I've merged in #329, and actually it's true by default, so it's opt out. I don't know if I'll change my mind later, but for now by default the highlights this plugin sets are default, so can be overridden. As I've said many times, it's still not the right thing to do to try and manipulate them because they are dynamic and can change if the config structure changes or the highlight generation logic changes. 

The help text now contains a list of the native nvim highlight groups that colorscheme authors should target instead/ensure are set, since those are what this plugin depends on.

I think it's fine to close this out now, unless there are some objections in the next day or so

**Comment by akinsho** [comment]
Closing this out since there are no comments or objections and highlight groups are default by default in dev soon to be main
